src/metabase/lib/datascript.cljc

```diff
@@ -0,0 +1,1361 @@
+    ;; Safe to assume this is a column ref, since subexpressions have no entity ref.
+    (entid? arg)        {:mbql/ref arg}
+    :else (throw (ex-info (str "clause-argument does not know what to do with " (pr-str arg)) {:arg arg}))))
+
+(defn- clause-arguments [args]
+  (vec (map-indexed (fn [i arg]
+                      (assoc (clause-argument arg)
+                             :mbql/series i
+                             :mbql/uuid   (str (random-uuid))))
+                    args)))
+
+(defn expr
+  "Shorthand for creating expression ASTs. This is like the named [[metabase.lib.expression/=]] etc., but I don't want
+  to wrestle with the name collisions in this big playground namespace.
+
+  The `args` can be DataScript Entities, `[:metadata.field/id 12]` refs, maps (typically nested clauses), or literal
+  numbers and strings. (Note that they can't be plain numeric entity IDs - those will get interpreted as literal
+  numbers!)
+
+  Returns the `:mbql.clause/*` entity in a transactable form."
+  [op & args]
+  {:mbql/uuid            (str (random-uuid))
+   :mbql.clause/operator op
+   :mbql.clause/argument (clause-arguments args)})
+
+;; ### Aggregations
+(defn- aggregation-clause
+  "Builder for an aggregation. Takes 1 to 3 args:
+
+  - `(aggregation-clause :count)`
+  - `(aggregation-clause :sum [:metadata.field/id 12])`
+  - `(aggregation-clause :sum-where [:metadata.field/id 12] (expr := [:metadata.field/id 9] 7))`
+      - `column` can be nil: `(aggregation-clause :count-where nil (expr := [:metadata.field/id 9] 7))`"
+  ([op] (aggregation-clause op nil nil))
+  ([op column] (aggregation-clause op column nil))
+  ([op column filter-expr]
+   (cond-> {:mbql.aggregation/operator (keyword op)
+            :mbql/uuid                 (str (random-uuid))}
+     column      (assoc :mbql.aggregation/column column)
+     filter-expr (assoc :mbql.aggregation/filter filter-expr))))
+
+(defn agg-count
+  "Returns a `:count` aggregation clause."
+  []
+  (aggregation-clause :count))
+
+(defn agg-sum
+  "Returns a `:sum` aggregation clause."
+  [column]
+  (aggregation-clause :sum column))
+
+(defn agg-min
+  "Returns a `:min` aggregation clause."
+  [column]
+  (aggregation-clause :min column))
+
+(defn agg-max
+  "Returns a `:max` aggregation clause."
+  [column]
+  (aggregation-clause :max column))
+
+(defn agg-sum-where
+  "Returns a `:sum-where` aggregation clause."
+  [column filter-expr]
+  (aggregation-clause :sum-where column filter-expr))
+
+(defn agg-count-where
+  "Returns a `:count-where` aggregation clause."
+  [filter-expr]
+  (aggregation-clause :count-where nil filter-expr))
+
+(defn- stage-aggregations [stage]
+  (->> stage
+       :mbql.stage/aggregation
+       (sort-by :mbql/series)
+       vec))
+
+(defn aggregations
+  "Returns the (ordered) list of aggregations on the given stage, as a vector of Entities."
+  [query-entity stage-number]
+  (stage-aggregations (query-stage query-entity stage-number)))
+
+(defn aggregate
```

Comment:
Ah, I meant a single one per `:mbql/index`. I'm actually rewriting it into using `max` over the `:mbql/index` values, so that `remove-clause` can just leave gaps.

The lookup can actually be written as a Datalog query and then DS should take an optimal approach to evaluating it. But there's no magic to querying in Datomic's model; it uses the same indexes as the smart maps.

It's not worth the overhead for so small a set of entities, but we can `:db/index true` an attribute to store it in AVET. (See the Datomic docs about the [four indexes](https://docs.datomic.com/query/indexes.html) if you're curious how that works. Note that DataScript doesn't use VAET; AVET is just as good for reverse ref lookup with all data in one partition.)